File: processor/src/test/resources/approvals/jackson/LazyvalJacksonModule.java
package test;

import java.time.LocalDate;
import javax.annotation.processing.Generated;
import scenarios.java.Ids;
import scenarios.java.Isbn;
import scenarios.java.OrderDate;
import scenarios.java.Quantity;
import tools.jackson.core.JsonGenerator;
import tools.jackson.core.JsonParser;
import tools.jackson.core.Version;
import tools.jackson.databind.DeserializationContext;
import tools.jackson.databind.JacksonModule;
import tools.jackson.databind.SerializationContext;
import tools.jackson.databind.ValueDeserializer;
import tools.jackson.databind.ValueSerializer;
import tools.jackson.databind.deser.std.StdDeserializer;
import tools.jackson.databind.module.SimpleDeserializers;
import tools.jackson.databind.module.SimpleModule;
import tools.jackson.databind.module.SimpleSerializers;
import tools.jackson.databind.ser.std.StdSerializer;

@Generated("com.qualityminds.lazyval.processor.internal.codegen.jackson.Jackson3Generator")
public class LazyvalJacksonModule extends SimpleModule {
  public LazyvalJacksonModule() {
    super("LazyvalJacksonModule", Version.unknownVersion());
  }

  @Override
  public void setupModule(JacksonModule.SetupContext context) {
    super.setupModule(context);
    SimpleSerializers sers = new SimpleSerializers();
    SimpleDeserializers desers = new SimpleDeserializers();
    sers.addSerializer(Ids.ProductId.class, IdsProductIdSerializer.INSTANCE);
    sers.addSerializer(Isbn.class, IsbnSerializer.INSTANCE);
    sers.addSerializer(OrderDate.class, OrderDateSerializer.INSTANCE);
    sers.addSerializer(Quantity.class, QuantitySerializer.INSTANCE);
    desers.addDeserializer(Ids.ProductId.class, IdsProductIdDeserializer.INSTANCE);
    desers.addDeserializer(Isbn.class, IsbnDeserializer.INSTANCE);
    desers.addDeserializer(OrderDate.class, OrderDateDeserializer.INSTANCE);
    desers.addDeserializer(Quantity.class, QuantityDeserializer.INSTANCE);
    context.addSerializers(sers);
    context.addDeserializers(desers);
  }

  static class IdsProductIdSerializer extends StdSerializer<Ids.ProductId> {
    static final IdsProductIdSerializer INSTANCE = new IdsProductIdSerializer();

    protected IdsProductIdSerializer() {
      super(Ids.ProductId.class);
    }

    @Override
    public void serialize(Ids.ProductId value, JsonGenerator gen, SerializationContext provider) {
      gen.writeString(value.value());
    }
  }

  static class IsbnSerializer extends StdSerializer<Isbn> {
    static final IsbnSerializer INSTANCE = new IsbnSerializer();

    protected IsbnSerializer() {
      super(Isbn.class);
    }

    @Override
    public void serialize(Isbn value, JsonGenerator gen, SerializationContext provider) {
      gen.writeString(value.getValue());
    }
  }

  static class OrderDateSerializer extends StdSerializer<OrderDate> {
    static final OrderDateSerializer INSTANCE = new OrderDateSerializer();

    private ValueSerializer<Object> innerSerializer;

    protected OrderDateSerializer() {
      super(OrderDate.class);
    }

    @Override
    public void serialize(OrderDate value, JsonGenerator gen, SerializationContext provider) {
      ValueSerializer<Object> ser = this.innerSerializer;
      if (ser == null) {
        ser = provider.findValueSerializer(LocalDate.class);
        this.innerSerializer = ser;
      }
      ser.serialize(value.value(), gen, provider);
    }
  }

  static class QuantitySerializer extends StdSerializer<Quantity> {
    static final QuantitySerializer INSTANCE = new QuantitySerializer();

    protected QuantitySerializer() {
      super(Quantity.class);
    }

    @Override
    public void serialize(Quantity value, JsonGenerator gen, SerializationContext provider) {
      gen.writeNumber(value.value());
    }
  }

  static class IdsProductIdDeserializer extends StdDeserializer<Ids.ProductId> {
    static final IdsProductIdDeserializer INSTANCE = new IdsProductIdDeserializer();

    protected IdsProductIdDeserializer() {
      super(Ids.ProductId.class);
    }

    @Override
    public Ids.ProductId deserialize(JsonParser p, DeserializationContext ctxt) {
      var value = p.getValueAsString();
      return Ids.ProductId.of(value);
    }
  }

  static class IsbnDeserializer extends StdDeserializer<Isbn> {
    static final IsbnDeserializer INSTANCE = new IsbnDeserializer();

    protected IsbnDeserializer() {
      super(Isbn.class);
    }

    @Override
    public Isbn deserialize(JsonParser p, DeserializationContext ctxt) {
      var value = p.getValueAsString();
      return Isbn.parse(value);
    }
  }

  static class OrderDateDeserializer extends StdDeserializer<OrderDate> {
    static final OrderDateDeserializer INSTANCE = new OrderDateDeserializer();

    private ValueDeserializer<Object> innerDeserializer;

    protected OrderDateDeserializer() {
      super(OrderDate.class);
    }

    @Override
    public OrderDate deserialize(JsonParser p, DeserializationContext ctxt) {
      ValueDeserializer<Object> deser = this.innerDeserializer;
      if (deser == null) {
        deser = ctxt.findContextualValueDeserializer(ctxt.constructType(LocalDate.class), null);
        this.innerDeserializer = deser;
      }
      var value = (LocalDate) deser.deserialize(p, ctxt);
      return new OrderDate(value);
    }
  }

  static class QuantityDeserializer extends StdDeserializer<Quantity> {
    static final QuantityDeserializer INSTANCE = new QuantityDeserializer();

    protected QuantityDeserializer() {
      super(Quantity.class);
    }

    @Override
    public Quantity deserialize(JsonParser p, DeserializationContext ctxt) {
      var value = p.getValueAsInt();
      return new Quantity(value);
    }
  }
}

File: ksp/src/test/resources/approvals/jackson/LazyvalJacksonModule.kt
package test

import java.time.LocalDate
import kotlin.Any
import scenarios.kotlin.Ids
import scenarios.kotlin.Isbn
import scenarios.kotlin.NullableQuantity
import scenarios.kotlin.OrderDate
import scenarios.kotlin.Quantity
import tools.jackson.core.JsonGenerator
import tools.jackson.core.JsonParser
import tools.jackson.core.Version
import tools.jackson.databind.DeserializationContext
import tools.jackson.databind.JacksonModule
import tools.jackson.databind.SerializationContext
import tools.jackson.databind.ValueDeserializer
import tools.jackson.databind.ValueSerializer
import tools.jackson.databind.deser.std.StdDeserializer
import tools.jackson.databind.module.SimpleDeserializers
import tools.jackson.databind.module.SimpleModule
import tools.jackson.databind.module.SimpleSerializers
import tools.jackson.databind.ser.std.StdSerializer

public class LazyvalJacksonModule : SimpleModule("LazyvalJacksonModule", Version.unknownVersion()) {
  override fun setupModule(context: JacksonModule.SetupContext) {
    super.setupModule(context)
    val sers = SimpleSerializers()
    val desers = SimpleDeserializers()
    sers.addSerializer(Ids.ProductId::class.java, idsProductIdSerializer)
    sers.addSerializer(Isbn::class.java, isbnSerializer)
    sers.addSerializer(NullableQuantity::class.java, nullableQuantitySerializer)
    sers.addSerializer(OrderDate::class.java, orderDateSerializer)
    sers.addSerializer(Quantity::class.java, quantitySerializer)
    desers.addDeserializer(Ids.ProductId::class.java, idsProductIdDeserializer)
    desers.addDeserializer(Isbn::class.java, isbnDeserializer)
    desers.addDeserializer(NullableQuantity::class.java, nullableQuantityDeserializer)
    desers.addDeserializer(OrderDate::class.java, orderDateDeserializer)
    desers.addDeserializer(Quantity::class.java, quantityDeserializer)
    context.addSerializers(sers)
    context.addDeserializers(desers)
  }

  public companion object {
    private val idsProductIdSerializer: IdsProductIdSerializer = IdsProductIdSerializer()

    private val isbnSerializer: IsbnSerializer = IsbnSerializer()

    private val nullableQuantitySerializer: NullableQuantitySerializer =
        NullableQuantitySerializer()

    private val orderDateSerializer: OrderDateSerializer = OrderDateSerializer()

    private val quantitySerializer: QuantitySerializer = QuantitySerializer()

    private val idsProductIdDeserializer: IdsProductIdDeserializer = IdsProductIdDeserializer()

    private val isbnDeserializer: IsbnDeserializer = IsbnDeserializer()

    private val nullableQuantityDeserializer: NullableQuantityDeserializer =
        NullableQuantityDeserializer()

    private val orderDateDeserializer: OrderDateDeserializer = OrderDateDeserializer()

    private val quantityDeserializer: QuantityDeserializer = QuantityDeserializer()
  }

  private class IdsProductIdSerializer : StdSerializer<Ids.ProductId>(Ids.ProductId::class.java) {
    override fun serialize(
      `value`: Ids.ProductId,
      gen: JsonGenerator,
      ctx: SerializationContext,
    ) {
      gen.writeString(value.value)
    }
  }

  private class IsbnSerializer : StdSerializer<Isbn>(Isbn::class.java) {
    override fun serialize(
      `value`: Isbn,
      gen: JsonGenerator,
      ctx: SerializationContext,
    ) {
      gen.writeString(value.value)
    }
  }

  private class NullableQuantitySerializer : StdSerializer<NullableQuantity>(NullableQuantity::class.java) {
    override fun serialize(
      `value`: NullableQuantity,
      gen: JsonGenerator,
      ctx: SerializationContext,
    ) {
      gen.writeNumber(value.value)
    }
  }

  private class OrderDateSerializer : StdSerializer<OrderDate>(OrderDate::class.java) {
    private var innerSerializer: ValueSerializer<Any>? = null

    override fun serialize(
      `value`: OrderDate,
      gen: JsonGenerator,
      ctx: SerializationContext,
    ) {
      val ser = innerSerializer ?: ctx.findValueSerializer(LocalDate::class.java).also { innerSerializer = it }
      ser.serialize(value.date, gen, ctx)
    }
  }

  private class QuantitySerializer : StdSerializer<Quantity>(Quantity::class.java) {
    override fun serialize(
      `value`: Quantity,
      gen: JsonGenerator,
      ctx: SerializationContext,
    ) {
      gen.writeNumber(value.value)
    }
  }

  private class IdsProductIdDeserializer : StdDeserializer<Ids.ProductId>(Ids.ProductId::class.java) {
    override fun deserialize(p: JsonParser, ctx: DeserializationContext): Ids.ProductId {
      val value = p.valueAsString
      return Ids.ProductId.of(value)
    }
  }

  private class IsbnDeserializer : StdDeserializer<Isbn>(Isbn::class.java) {
    override fun deserialize(p: JsonParser, ctx: DeserializationContext): Isbn {
      val value = p.valueAsString
      return Isbn.parse(value)
    }
  }

  private class NullableQuantityDeserializer : StdDeserializer<NullableQuantity?>(NullableQuantity::class.java) {
    override fun deserialize(p: JsonParser, ctx: DeserializationContext): NullableQuantity? {
      val value = p.valueAsInt
      return NullableQuantity.ofNullable(value)
    }
  }

  private class OrderDateDeserializer : StdDeserializer<OrderDate>(OrderDate::class.java) {
    private var innerDeserializer: ValueDeserializer<Any>? = null

    override fun deserialize(p: JsonParser, ctx: DeserializationContext): OrderDate {
      val deser = innerDeserializer ?: ctx.findContextualValueDeserializer(ctx.constructType(LocalDate::class.java), null).also { innerDeserializer = it }
      val value = deser.deserialize(p, ctx) as LocalDate
      return OrderDate(value)
    }
  }

  private class QuantityDeserializer : StdDeserializer<Quantity>(Quantity::class.java) {
    override fun deserialize(p: JsonParser, ctx: DeserializationContext): Quantity {
      val value = p.valueAsInt
      return Quantity(value)
    }
  }
}
